plugins/regex_revalidate/regex_revalidate.c

```diff
@@ -343,35 +344,33 @@ free_handler(TSCont cont, TSEvent event, void *edata)
 static int
 config_handler(TSCont cont, TSEvent event, void *edata)
 {
-  plugin_state_t *pstate;
-  invalidate_t *i, *iptr;
-  TSCont free_cont;
-  bool updated;
-  TSMutex mutex;
-
-  mutex = TSContMutexGet(cont);
+  TSMutex mutex = TSContMutexGet(cont);
   TSMutexLock(mutex);
 
   TSDebug(LOG_PREFIX, "In config Handler");
-  pstate = (plugin_state_t *)TSContDataGet(cont);
-  i      = copy_config(pstate->invalidate_list);
+  plugin_state_t *const pstate = (plugin_state_t *)TSContDataGet(cont);
+
+  invalidate_t *listold = atomic_load(&(pstate->invalidate_list));
+  invalidate_t *listnew = copy_config(listold);
 
-  updated = prune_config(&i);
-  updated = load_config(pstate, &i) || updated;
+  bool const loaded = load_config(pstate, &listnew);
+  bool const pruned = prune_config(&listnew);
+
+  bool const updated = pruned || loaded;
 
   if (updated) {
-    list_config(pstate, i);
-    iptr = __sync_val_compare_and_swap(&(pstate->invalidate_list), pstate->invalidate_list, i);
+    list_config(pstate, listnew);
+    listold = atomic_exchange(&(pstate->invalidate_list), listnew);
```

Comment:
Change `CAS` to `exchange` seems a bit worrisome, but I can't figure out what the original `CAS` was trying to do. The point would be to protect against race conditions where multiple configuration updates are happening. This could be for the update or because an intermediate value was retrieved somewhere else and it ends up being destroyed before use.

Reply:
I switched the atomic_exchange over to atomic_compare_exchange_strong to check if the original config pointer may have been switched out from under the config_handler.

I'm not sure this is even possible however, or if this PR with the extra atomics checks even does anything to help with thread safety.

Reply:
Just removed the CAS and now atomically store the new config and have the transaction atomically fetch the current config.  Theoretically this should be sufficient.